## Machine Learning Record Mining

Project to create a pipeline that uses GeoDeepDive's output to find Unaquired Sites for Neotoma.

Using NLP parsed text and a Data Science approach, identify whether a paper is suitable for Neotoma and detect features such as 'Site Name', 'Location', 'Age Span' and 'Site Descriptions'.

In [3]:
# Loading libraries

import numpy as np
import pandas as pd
import csv
import psycopg2
import re

# Load Postgres Server
from src.config import config

In [4]:
# Options for DF display
#pd.set_option('display.max_colwidth', 50)
#pd.set_option('display.max_rows', 10)

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', -1)

## Loading and viewing the Data

### Loading NLP Sentences

In [5]:
# Connect to PostgreSQL server from terminal:
# pg_ctl -D PSQL_Data -l logfile start

try:
    params = config()
    # Connect to the PostgreSQL database
    conn = psycopg2.connect(**params)
    # Create a new cursor
    cur = conn.cursor()
    nlp_sentences = pd.read_sql_query('''SELECT * FROM sentences;''', conn)
    # Close the cursor and connection to so the server can allocate
    # bandwidth to other requests
    cur.close()
    conn.close()
    print('Data obtained from PostgreSQL')
    
# If no SQL db, load from a file
except:
    header_list = ["_gddid", "sentid", "wordidx", "words", "part_of_speech", "special_class", 
               "lemmas", "word_type", "word_modified"]
    nlp_sentences = pd.read_csv("../Do_not_commit_data/sentences_nlp352", sep='\t', names = header_list)
    nlp_sentences = nlp_sentences.replace('"', '', regex = True)\
                                 .replace('\{', '', regex = True)\
                                 .replace('}', '', regex = True)\
                                 .replace(',', ',', regex = True)
    nlp_sentences['wordidx']= nlp_sentences['wordidx'].str.split(",")
    nlp_sentences['words']= nlp_sentences['words'].str.split(",")
    nlp_sentences['poses']= nlp_sentences['poses'].str.split(",")
    nlp_sentences['ners']= nlp_sentences['ners'].str.split(",")
    nlp_sentences['lemmas']= nlp_sentences['lemmas'].str.split(",")
    nlp_sentences['dep_paths']= nlp_sentences['dep_paths'].str.split(",")
    nlp_sentences['dep_parents']= nlp_sentences['dep_parents'].str.split(",")
    print('Data obtained from text file')

Data obtained from PostgreSQL


In [6]:
nlp_sentences.head(3)

_gddid  sentid  \
0  54b43266e138239d8684efed  1        
1  54b43266e138239d8684efed  2        
2  54b43266e138239d8684efed  3        

                                                                                                                                                                                                                                                                                                                                                                                                         wordidx  \
0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, ...]   
1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]                                                                                                                                                                                                                                                                                                                                                                         
2  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]                                                                                                                                                                                                                                                                                                                 

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       words  \
0  [Available, online, at, www.sciencedirect.com, Quaternary, Research, 69, -LRB-, 2008, -RRB-, 263, --, 275, www.elsevier.com/locate/yqres, Development, of, the, mixed, conifer, forest, in, northern, New, Mexico, and, its, relationship, to, Holocene, environmental, change, R., Scott, Anderson, a, ,, b, ,, ,, Renata, B., Jass, b, ,1, ,, Jaime, L., Toney, b, ,2, ,, Craig, D., Allen, c, ,, Luz, M., Cisneros-Dozal, d, ,, Marcey, Hess, d, ,, Jeff, Heikoop, d, ,, Julianna, Fessenden, d, a, Center, for, Environmental, Sciences, &, Education, ,, Box, 5694, ,, Northern, Arizona, University, ,, Flagstaff, ,, AZ, 86011, ,, USA, b, Quaternary, Sciences, Program, &, Bilby, Research, ...]   
1  [The, Chihuahueños, Bog, record, extends, to, over, 15,000, cal, yr, BP, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
2  [An, Artemisia, steppe, ,, then, an, open, Picea, woodland, grew, around, a, small, pond, until, ca., 11,700, cal, yr, BP, when, Pinus, ponderosa, became, established, .]          

In [7]:
# Running this takes forever. Do not run unless necessary
# nlp_sentences.describe(include = 'all')

In [8]:
avg_sentences = nlp_sentences.groupby('_gddid')['_gddid','sentid'].transform('max')
avg_sentences.drop_duplicates(subset='_gddid', keep='first', inplace=True)
avg_sentences.describe()

sentid
count  218.000000 
mean   513.600917 
std    232.460891 
min    2.000000   
25%    397.250000 
50%    504.000000 
75%    638.250000 
max    1330.000000

From the Summary Statistics, despite having 111,965 sentences, there are only 218 unique articles (`_gddid`). The article with most sentences, had 1330 sentences; that's the max number found in `sentid`.  
In order to know the average of sentences, I grouped by `_gddid` and then, obtained the number of sentences per article. After that, I just got the mean through a Summary of statistics to know that on average, each article has 514 sentences.

### Loading Bibliography Data

In [9]:
import json
with open('../Do_not_commit_data/bibjson', 'r') as f:
    bib_dict = json.load(f)

In [10]:
# Normalizing data so that we have access to the 'identifier'

# TODO Load into SQL server and connect through SQL
bibliography = pd.io.json.json_normalize(bib_dict,'identifier',['publisher', 'title', ['journal','name', 'name'], ['author'], 'year', 'number', 'volumne', ['link'], '_gddid', 'type', 'pages'], record_prefix='_', errors='ignore')

bibliography['link'] = bibliography['link'].astype(str)

url = bibliography['link'].str.split(", ", expand = True)

bibliography['link_url'] = url[0]
bibliography['link_type'] = url[1]

bibliography['link_url'] = bibliography['link_url'].replace("\[{'url': '", "", regex = True)\
                                                   .replace("'", "", regex = True)

bibliography['link_type'] = bibliography['link_type'].replace("'type': '", "", regex = True)\
                                                     .replace("'}]", "", regex = True)

bibliography['author'] = bibliography['author'].astype(str)
bibliography['author'] = bibliography['author'].replace("\[{'name': '", "", regex = True)\
                                               .replace("{'name': '", "", regex = True)\
                                               .replace("'},", ";", regex = True)\
                                               .replace("'}]", "", regex = True)
bibliography = bibliography[['_type', '_id', 'publisher', 'title', 'journal.name.name',	'author', 'year', 'number', 'volumne', '_gddid', 'type', 'pages', 'link_url', 'link_type']]

bibliography.head(3)

_type                            _id                    publisher  \
0  doi   10.1016/S0277-3791(99)00007-4  Elsevier                      
1  doi   10.1139/b92-002                Canadian Science Publishing   
2  doi   10.1016/0031-0182(80)90043-7   Elsevier                      

                                                                                                                                                                         title  \
0  Palaeoclimate, chronology and vegetation history of the Weichselian Lateglacial: comparative analysis of data from three cores at Lago Grande di Monticchio, southern Italy   
1  Holocene history of forest trees on the Bruce Peninsula, southern Ontario                                                                                                     
2  Glacial sequence and environmental history in the Sierra Nevada del cocuy (Colombia)                                                                                          

                                   journal.name.name  \
0  Quaternary Science Reviews                          
1  Canadian Journal of Botany                          
2  Palaeogeography, Palaeoclimatology, Palaeoecology   

                                                        author  year number  \
0  Huntley, B.; Watts, W.A.; Allen, J.R.M.; Zolitschka, B.      1999  7       
1  Bennett, K. D.                                               1992  1       
2  Van Der Hammen, T.; Barelds, J.; De Jong, H.; De Veer, A.A.  1980  NaN     

  volumne                    _gddid     type     pages  \
0  NaN     550453fde1382326932d85f7  article  945--960   
1  NaN     578b5aabcf58f1587003efba  article  6--18      
2  NaN     54b43244e138239d868493cd  article  247--340   

                                                             link_url  \
0  http://www.sciencedirect.com/science/article/pii/S0277379199000074   
1  http://www.nrcresearchpress.com/doi/abs/10.1139/b92-002              
2  http://www.sciencedirect.com/science/article/pii/0031018280900437    

   link_type  
0  publisher  
1  publisher  
2  publisher

In [11]:
bibliography.describe(include = 'all')

_type                           _id publisher  \
count   221   221                           221        
unique  1     221                           6          
top     doi   10.1016/j.foreco.2007.10.008  Elsevier   
freq    221   1                             138        

                                                      title  \
count   221                                                   
unique  221                                                   
top     Holocene climate variability in northernmost Europe   
freq    1                                                     

          journal.name.name          author  year number volumne  \
count   221                  221             221   183    0        
unique  41                   213             47    28     0        
top     Quaternary Research  Markgraf, Vera  2011  3      NaN      
freq    44                   3               14    40     NaN      

                          _gddid     type  pages  \
count   221                       221      216     
unique  221                       1        214     
top     5714568acf58f125068458dc  article  1--13   
freq    1                         221      2       

                                                     link_url  link_type  
count   221                                                    220        
unique  221                                                    1          
top     http://doi.wiley.com/10.1111/j.1502-3885.2010.00159.x  publisher  
freq    1                                                      220

## EDA

Reviewing our data includes skimming through some papers online and seeing if the data is consistent with our NLP Sentences dataframe.

From there, we can also visualize what we would like our model to predict: 'Location', 'Site Name', 'Age Span', and 'Site Description' from a "Human perspective".

### Finding Latitude and Longitud via REGEX 

After reviewing some PDF's and viewing the data parsing, I noticed some possible patterns for coordinates.

<img src="figures/img/00_snippet_of_pdf.png" alt="Drawing" style="width: 300px;"/>

Which translates to our pandas dataframe as:

![alt text](figures/img/01_snippet_in_sql.png "PandasVersion")

As a first problem, we can notice that the degrees symbol `°` was translated as a `3`. On some other occassions, it is confused as an `o` or a middle point `◦`. For `'`, this symbol got confused by the word `cents`.

I am still going to use normal REGEX to find coordinates in the most intuitive way. 

A coordinate for longitude is a number ranging from -180° to 180°. It has three components, `°, ',"` and it covers the position in E or W.

Latitude goes from -90° to 90°, has the same three components and goes from S to N.

Using REGEX, and assuming data consistency, the following code should be able to extract coordinates. 

In [12]:
dms_regex = r"([-]?\d{1,3}[°|′|\'|,]{0,3}\d{1,2}[,|′|\']{1,3}\d{0,2}[,|′|\']{1,4}[NESWnesw]?[\s|,|\']+?[-]?\d{1,3}[°|,|′|\']+\d{1,2}[,|′|\']+\d{0,2}[,|′|\'][,|′|\']{0,4}[NESWnesw]?)"

dd_regex = "[\\{,][-]?[1]?[0-9]{1,2}\\.[0-9]{1,}[,]?[NESWnesw],"

In [13]:
# TODO Find REGEX of couples of numbers (123, 234)
nlp_sentences['words_as_string'] = nlp_sentences['words'].apply(lambda x: ','.join(map(str, x)))
nlp_sentences['dms_regex'] = nlp_sentences['words_as_string'].str.findall(dms_regex)


#nlp_sentences[['dms_regex', 'words']]


#nlp_sentences['dms_regex'] = nlp_sentences['dms_regex'].astype(str)
#nlp_sentences['dms_regex'] = nlp_sentences['dms_regex'].replace(",", ";", regex = True)#\
                                                       #.replace(",", "", regex = True)

nlp_sentences[['words_as_string', 'words', 'dms_regex']].head(3)

words_as_string  \
0  Available,online,at,www.sciencedirect.com,Quaternary,Research,69,-LRB-,2008,-RRB-,263,--,275,www.elsevier.com/locate/yqres,Development,of,the,mixed,conifer,forest,in,northern,New,Mexico,and,its,relationship,to,Holocene,environmental,change,R.,Scott,Anderson,a,,,b,,,,,Renata,B.,Jass,b,,1,,,Jaime,L.,Toney,b,,2,,,Craig,D.,Allen,c,,,Luz,M.,Cisneros-Dozal,d,,,Marcey,Hess,d,,,Jeff,Heikoop,d,,,Julianna,Fessenden,d,a,Center,for,Environmental,Sciences,&,Education,,,Box,5694,,,Northern,Arizona,University,,,Flagstaff,,,AZ,86011,,,USA,b,Quaternary,Sciences,Program,&,Bilby,Research,Center,,,Box,6013,,,Northern,Arizona,University,,,Flagstaff,,,AZ,86011,,,USA,c,U.S.,Geological,Survey,,,Jemez,Mountains,Field,Station,,,Bandelier,National,Monument,,,HCR-1,,,Box,1,,,#,15,,,Los,Alamos,,,NM,87544,,,USA,d,Earth,&,Environmental,Sciences,Division,,,Hydrology,,,Geochemistry,&,Geology,Group,,,EES-6,,,MS-D462,,,Los,Alamos,National,Laboratory,,,Los,Alamos,,,NM,87545,,,USA,Received,28,April,2007,Available,online,29,January,2008,Abstract,Chihuahueños,Bog,-LRB-,2925,m,-RRB-,in,the,Jemez,Mountains,of,northern,New,Mexico,contains,one,of,the,few,records,of,late-glacial,and,postglacial,development,of,the,mixed,conifer,forest,in,southwestern,North,America,.   
1  The,Chihuahueños,Bog,record,extends,to,over,15,000,cal,yr,BP,.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
2  An,Artemisia,steppe,,,then,an,open,Picea,woodland,grew,around,a,small,pond,until,ca.,11,700,cal,yr,BP,when,Pinus,ponderosa,became,established,.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             

                                                                                                                                                                                                                                            

In [14]:
#nlp_sentences[['words_as_string', 'words', 'dms_regex']].describe()

The `dms_regex` only returned 95 unique values. That's to say that from the 111965 sentences, only 95 possess something similar to the regex we are looking for.  

Let's see an example of a regex straction as a sanity check that we are getting the values that we need. 

In [15]:
italy_example = nlp_sentences[nlp_sentences['_gddid'] == '550453fde1382326932d85f7']
italy_example = italy_example[italy_example['sentid'] == 10]
italy_example[['words', 'words_as_string', 'dms_regex']]

words  \
94875  [Introduction, Laghi, di, Monticchio, are, two, maar, lakes, located, in, the, explosion, crater, that, lies, west, of, the, peak, of, Monte, Vulture, ,, near, Mel, '', in, the, Basilicata, region, of, southern, Italy, -LRB-, 403, 56, 40, N, ,, 153, 36, 48, E, -RRB-, -LRB-, Fig., 1, -RRB-, .]   

                                                                                                                                                                                                                                           words_as_string  \
94875  Introduction,Laghi,di,Monticchio,are,two,maar,lakes,located,in,the,explosion,crater,that,lies,west,of,the,peak,of,Monte,Vulture,,,near,Mel,'',in,the,Basilicata,region,of,southern,Italy,-LRB-,403,56,40,N,,,153,36,48,E,-RRB-,-LRB-,Fig.,1,-RRB-,.   

                         dms_regex  
94875  [403,56,40,N,,,153,36,48,E]

### Problems with the REGEX approach

As seen before, extracting information using REGEX is very complicated. We would require a lot of queries that would retrieve each particular case.

A possible alternative could be using NER, Named Entity Recognition.

In [16]:
def order_article(article_id):
    '''
    Function to find an article by its gddid in the NLP sentences and have it displayed in order
    
    Keyword arguments:
    article_id -- gddid
    
    Returns:
    article ordered by sentence index
    '''
    article = nlp_sentences[nlp_sentences['_gddid'] == article_id]
    return article[['sentid', 'words']].sort_values(by = 'sentid')

Article 01  
Skimmed info: 
 * `Article Name:` Palaeoclimate_chronology_and_vegetation_history_of_the_Weichselian_Lateglacial
 * `Link`: http://www.sciencedirect.com/science/article/pii/S0277379199000074  
 * `Site Name`:  Laghi di Monticchio
 * `Location`:  40° 56' 40" N, 15° 36' 48" E
 * `Age Span`: 
 * `Site Descriptions`: The record from Lago Grande di Monticchio now adds macrofossil evidence of the presence of relatively boreal tree taxa during the Lateglacial interstadial and provides a chronology based upon annually laminated sediments.

In [17]:
# Article 01 
order_article('550453fde1382326932d85f7')

sentid  \
94865  1        
94866  2        
94867  3        
94868  4        
94870  5        
94871  6        
94872  7        
94873  8        
94874  9        
94875  10       
94877  11       
94878  12       
94879  13       
94880  14       
94881  15       
94882  16       
94883  17       
94884  18       
94885  19       
94887  20       
94888  21       
94889  22       
94890  23       
94891  24       
94892  25       
94893  26       
94894  27       
94895  28       
94896  29       
94897  30       
94898  31       
94899  32       
94785  33       
94901  34       
94902  35       
94903  36       
94904  37       
94905  38       
94906  39       
94876  40       
94907  41       
94908  42       
94909  43       
94910  44       
94911  45       
94912  46       
94914  47       
94915  48       
94916  49       
94917  50       
94919  51       
94920  52       
94921  53       
94922  54       
94918  55       
94923  56       
94913  57       
94924  58       
94925  59       
94926  60       
94927  61       
94928  62       
94929  63       
94930  64       
94931  65       
94932  66       
94933  67       
94934  68       
94935  69       
94937  70       
94938  71       
94939  72       
94940  73       
94941  74       
94943  75       
94944  76       
94945  77       
94946  78       
94947  79       
94948  80       
94949  81       
94950  82       
94951  83       
94952  84       
94954  85       
94955  86       
94956  87       
94957  88       
94958  89       
94959  90       
94960  91       
94961  92       
94962  93       
94963  94       
94964  95       
94965  96       
94966  97       
94967  98       
94968  99       
94970  100      
94971  101      
94972  102      
94973  103      
94974  104      
94976  105      
94977  106      
94978  107      
94979  108      
94980  109      
94982  110      
94983  111      
94984  112      
94985  113      
94986  114      
94987  115      
94988  116      
94989  117      
94990  118      
94991  119      
94992  120      
94981  121      
94969  122      
94993  123      
94994  124      
94995  125      
94996  126      
94997  127      
94999  128      
95000  129      
95001  130      
95002  131      
95003  132      
95005  133      
95006  134      
95007  135      
95008  136      
95009  137      
95010  138      
95012  139      
95013  140      
95014  141      
95015  142      
95016  143      
95017  144      
95018  145      
95019  146      
95020  147      
95021  148      
95022  149      
94936  150      
94942  151      
95023  152      
95024  153      
95025  154      
95026  155      
95028  156      
95029  157      
95030  158      
95031  159      
95032  160      
95033  161      
94975  162      
95027  163      
95035  164      
95036  165      
95037  166      
95038  167      
95039  168      
95041  169      
95042  170      
95043  171      
95044  172      
95045  173      
95047  174      
95048  175      
95049  176      
95050  177      
95052  178      
95053  179      
95054  180      
95055  181      
95057  182      
95058  183      
95059  184      
95060  185      
95061  186      
95062  187      
95063  188      
95066  189      
95067  190      
95068  191      
95069  192      
95070  193      
95071  194      
95072  195      
95073  196      
95074  197      
95075  198      
95076  199      
95077  200      
95078  201      
95079  202      
95080  203      
95056  204      
95064  205      
95081  206      
95082  207      
95083  208      
95084  209      
95085  210      
95086  211      
95087  212      
95088  213      
95089  214      
95090  215      
95091  216      
95093  217      
95094  218      
95095  219      
95096  220      
95097  221      
95099  222      
95100  223      
95101  224      
95102  225      
95103  226      
95092  227      
95104  228      
95105  229      
95106  230      
95107  231      
95109  232      
95110  233      
95111  234      
95112  235  

Article 02  
Skimmed info:  
* `Article:` Development of the Mixed Connifer Forest in Northern New Mexico and its relationship to Holocene environmental change
* `Link`: http://www.sciencedirect.com/science/article/pii/S0033589407001512
* `Site Name`:  Chihuahuenos Bog 
* `Location`:  ??? No Given Coordinates 
* `Age Span`:   
* `Site Descriptions`:  Chihuahueños Bog is located within the mixed conifer forest and has affinities to the flora of the southern Rocky Mountains(Foxx and Tierney, 1985; Allen, 2004). Although the area immediately surrounding the bog has been heavily logged in recent years, nearly every conifer species growing in montaneforests of the Jemez mountains is represented either along theshore of the bog or in the surrounding uplands.  

In [18]:
# Article 2
order_article('54b43266e138239d8684efed')

sentid  \
0    1        
1    2        
2    3        
3    4        
4    5        
6    6        
7    7        
8    8        
9    9        
10   10       
11   11       
12   12       
14   13       
15   14       
16   15       
17   16       
18   17       
19   18       
21   19       
22   20       
23   21       
24   22       
26   23       
27   24       
28   25       
29   26       
30   27       
25   28       
31   29       
32   30       
33   31       
34   32       
36   33       
37   34       
38   35       
39   36       
41   37       
42   38       
43   39       
44   40       
45   41       
47   42       
48   43       
49   44       
50   45       
51   46       
53   47       
54   48       
55   49       
56   50       
57   51       
60   52       
61   53       
62   54       
63   55       
64   56       
66   57       
67   58       
68   59       
69   60       
70   61       
71   62       
72   63       
73   64       
74   65       
75   66       
76   67       
78   68       
79   69       
80   70       
81   71       
82   72       
84   73       
85   74       
86   75       
87   76       
88   77       
89   78       
77   79       
91   80       
92   81       
93   82       
94   83       
95   84       
97   85       
98   86       
99   87       
100  88       
101  89       
102  90       
104  91       
105  92       
106  93       
107  94       
108  95       
110  96       
111  97       
112  98       
113  99       
114  100      
115  101      
116  102      
117  103      
118  104      
119  105      
103  106      
121  107      
122  108      
123  109      
124  110      
125  111      
126  112      
127  113      
128  114      
129  115      
130  116      
131  117      
132  118      
133  119      
134  120      
135  121      
137  122      
138  123      
139  124      
140  125      
141  126      
142  127      
143  128      
145  129      
146  130      
147  131      
148  132      
149  133      
150  134      
152  135      
153  136      
154  137      
155  138      
156  139      
157  140      
158  141      
159  142      
160  143      
161  144      
162  145      
164  146      
165  147      
166  148      
167  149      
151  150      
169  151      
170  152      
171  153      
172  154      
163  155      
174  156      
175  157      
176  158      
177  159      
178  160      
179  161      
180  162      
40   163      
182  164      
183  165      
184  166      
185  167      
187  168      
188  169      
189  170      
190  171      
191  172      
193  173      
194  174      
195  175      
196  176      
199  177      
200  178      
201  179      
202  180      
204  181      
205  182      
206  183      
207  184      
197  185      
209  186      
210  187      
211  188      
212  189      
214  190      
215  191      
216  192      
217  193      
219  194      
220  195      
221  196      
222  197      
223  198      
213  199      
224  200      
225  201      
226  202      
227  203      
228  204      
96   205      
231  206      
232  207      
233  208      
234  209      
237  210      
238  211      
239  212      
240  213      
242  214      
243  215      
244  216      
245  217      
246  218      
248  219      
249  220      
250  221      
251  222      
252  223      
254  224      
255  225      
256  226      
257  227      
259  228      
260  229      
261  230      
262  231      
263  232      
265  233      
266  234      
267  235      
268  236      
270  237      
271  238      
272  239      
273  240      
275  241      
276  242      
277  243      
278  244      
280  245      
281  246      
282  247      
283  248      
284  249      
287  250      
288  251      
289  252      
290  253      
292  254      
293  255      
294  256      
295  257      
296  258      
235  259      
291  260      
298  261      
299  262      
300  263      
301  264      
302  265      
304  266      


Article 03  
Skimmed info:  
* `Link`: http://dx.doi.org/10.1130/g35541.1  # No Full access to article
* `Site Name`:  Cahokia region, Illinois, USA
* `Location`:  N38°42'00" - N38°42'00", W90°04'60" - W90°04'60" - Not found on text (yet).
* `Age Span`:   
* `Site Descriptions`: Record from Horseshoe Lake, an oxbow lake in the central Mississippi River valley that is adjacent to the Cahokia site (Illinois, USA), the largest prehistoric settlement north of Mexico.   

In [19]:
# Article 3
order_article('57c5b941cf58f1338eaddb5b')

sentid  \
67708  1        
67709  2        
67710  3        
67712  4        
67713  5        
67714  6        
67715  7        
67716  8        
67718  9        
67719  10       
67720  11       
67721  12       
67706  13       
67717  14       
67723  15       
67724  16       
67725  17       
67726  18       
67727  19       
67728  20       
67729  21       
67730  22       
67732  23       
67733  24       
67734  25       
67735  26       
67736  27       
67738  28       
67739  29       
67740  30       
67741  31       
67743  32       
67744  33       
67745  34       
67746  35       
67747  36       
67749  37       
67750  38       
67751  39       
67752  40       
67754  41       
67755  42       
67756  43       
67757  44       
67731  45       
67759  46       
67760  47       
67761  48       
67762  49       
67764  50       
67765  51       
67766  52       
67767  53       
67769  54       
67770  55       
67771  56       
67772  57       
67773  58       
67768  59       
67774  60       
67775  61       
67776  62       
67777  63       
67778  64       
67780  65       
67781  66       
67782  67       
67783  68       
67785  69       
67786  70       
67787  71       
67788  72       
67789  73       
67790  74       
67791  75       
67792  76       
67793  77       
67794  78       
67795  79       
67796  80       
67797  81       
67798  82       
67800  83       
67801  84       
67802  85       
67803  86       
67804  87       
67805  88       
67806  89       
67807  90       
67808  91       
67809  92       
67810  93       
67811  94       
67813  95       
67814  96       
67815  97       
67816  98       
67817  99       
67818  100      
67819  101      
67820  102      
67821  103      
67823  104      
67824  105      
67825  106      
67826  107      
67827  108      
67828  109      
67829  110      
67830  111      
67832  112      
67833  113      
67834  114      
67835  115      
67836  116      
67838  117      
67839  118      
67840  119      
67841  120      
67843  121      
67844  122      
67845  123      
67846  124      
67848  125      
67849  126      
67850  127      

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         words  
67708  [A, record, of, sustained, prehistoric, and, historic, land, use, from, the, Cahokia, region, ,, Illinois, ,, USA, Samuel, E., Munoz1, ,, Sissel, Schroeder2, ,, David, A., Fike3, ,, and, John, W., Williams1, 1Department, of, Geography, ,, University, of, Wisconsin, --, Madison, ,, 550, North, Park, Street, ,, Madison, ,, Wisconsin, 53706, ,, USA, 2Department, of, Anthropology, ,, University, of, Wisconsin, --, Madison, ,, 1180, Observatory, Drive, ,, Madison, ,, Wisconsin, 53706, ,, USA, 3Department, of, Earth, and, Planetary, Sciences, ,, Washington, University, ,, 1, Brookings, Drive, ,, Saint, Louis, ,, Missouri, 63130, ,, USA, ABSTRACT, In, eastern, North, America, ,, ...]                                                                                                                                                                                                            
67709  [Here, we, report, a, high-re

Article 04  
Skimmed info:  
* `Link:`http://www.tandfonline.com/doi/abs/10.1080/00173130902965157
* `Site Name`:  Begbunar
* `Location`:  42°09′ N, 22° 33′ E; 1750 ; in SQL file shows as: `42,°,09,cents,N,,,22,°,33,cents,E`
* `Age Span`:   
* `Site Descriptions`: Located in the central treeless zone of the Osogovo mountains, which are situated at the border between south-western Bulgaria and the north-eastern former Yugoslav Republic of Macedonia.

In [20]:
# Article 4
order_article('58d29193cf58f14928755ba5')

sentid  \
110659  1        
110660  2        
110661  3        
110662  4        
110665  5        
110666  6        
110667  7        
110668  8        
110669  9        
110663  10       
110670  11       
110671  12       
110672  13       
110673  14       
110674  15       
110676  16       
110677  17       
110678  18       
110679  19       
110680  20       
110683  21       
110684  22       
110685  23       
110686  24       
110687  25       
110688  26       
110689  27       
110675  28       
110690  29       
110691  30       
110692  31       
110693  32       
110694  33       
110695  34       
110696  35       
110698  36       
110699  37       
110700  38       
110701  39       
110702  40       
110704  41       
110705  42       
110706  43       
110707  44       
110708  45       
110709  46       
110710  47       
110711  48       
110712  49       
110713  50       
110714  51       
110718  52       
110719  53       
110720  54       
110721  55       
110722  56       
110723  57       
110724  58       
110681  59       
110682  60       
110703  61       
110715  62       
110716  63       
108814  64       
110725  65       
110726  66       
110727  67       
110728  68       
110729  69       
110730  70       
110731  71       
110732  72       
110733  73       
110734  74       
110735  75       
110736  76       
110737  77       
110738  78       
110739  79       
110740  80       
110741  81       
110742  82       
110743  83       
110744  84       

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      words  
110659  [Grana, ISSN, :, 0017-3134, -LRB-, Print, -RRB-, 1651-2049, -LRB-, Online, -RRB-, Journal, homepage, :, http://www.tandfonline.com/loi/sgra20, 6, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
110660  [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains, ,, southwest, Bulgaria, -RRB-, :, Four, millennia, of, vegetation, history, Maria, Lazarova, ,, Spassimir, Tonkov, ,, Ian, Snowball, &, Elena, Marinova, To, cite, this, article, :, Maria, Lazarova, ,, Spassimir, Tonkov, ,, Ian, Snowball, &, Elena, Marinova, -LRB-, 2009, -RRB-, 6, .]                                                                                                                                                                                                                                                                                                                                                                                                                                          
110661  [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains, ,, south-west, Bulgaria, -RRB-, :, Four, millennia, of, vegetation, history, ,, Grana, ,, 48:2, ,, 147-149, ,, DOI, :, 10.1

In [21]:
# Try to see the exact location. Wont show last few words.
pd.DataFrame(order_article('58d29193cf58f14928755ba5').iloc[4])

110665
sentid  5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
words   [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains, ,, south-west, Bulgaria, -RRB-, :, Four, millennia, of, vegetation, history, MARIAEPDB6.Begbunar, -LRB-, Bulgaria, -RRB-, LAZAROVA1, ,, SPASSIMIR, TONKOV2, ,, IAN, SNOWBALL3, &, ELENA, MARINOVA4, 1Institute, of, Botany, ,, Bulgarian, Academy, of, Sciences, ,, Sofia, ,, Bulgaria, ,, 2Laboratory, of, Palynology, ,, Department, of, Botany, ,, Faculty, of, Biology, ,, Sofia, University, ``, St., Kliment, Ohridski, '', ,, Sofia, ,, Bulgaria, ,, 3GeoBiosphere, Science, Centre, ,, Quaternary, Sciences, ,, Department, of, Geology, ,, Lund, University, ,, Lund, ,, Sweden, ,, 4Center, for, Archaeological, Sciences, ,, Katholieke, Universiteit, Leuven, ,, Leuven, ,, Belgium, Site, details, The, peat-bog, ...]

Article 05  
Skimmed info:  
* `Link`: http://dx.doi.org/10.1130/0016-7606(1997)109<1306:tfalqp>2.3.co
* `Site Name`:  Cottonwood III, Cottonwood IV, Red Lady, Red Well, Splains
* `Location`:  38°49'50' 106°24'45', 38°49'50' 106°24'45', 38°52'50' 107°2'30', 38°52'50' 107°2'30',38°53'40' 107°3'15', 38°52'28' 107°05', 38°50' 107°4'30', 38°50' 107°4'30', 38°52' 107°2'30', 38°45' 106°50', 38°45' 106°50'
* `Age Span`:   17000 - 400
* `Site Descriptions`:   The lower timberline or lower forest border is the lower elevational limit of a continuous forest belt. In the Rocky Mountains it forms a sharp contrast with shrublands or grasslands in valley bottoms. Trees may extend below the lower timberline in favorable habitats, along streams or on rocky outcrops.  

In [22]:
# Article 5
order_article('57928e07cf58f133d1c26609')

sentid  \
39606  1        
39607  2        
39608  3        
39609  4        
39611  5        
39612  6        
39613  7        
39614  8        
39616  9        
39617  10       
39618  11       
39619  12       
39620  13       
39615  14       
39621  15       
39622  16       
39623  17       
39624  18       
39625  19       
39626  20       
39627  21       
39628  22       
39629  23       
39630  24       
39631  25       
39632  26       
39633  27       
39634  28       
39635  29       
39637  30       
39638  31       
39639  32       
39640  33       
39642  34       
39643  35       
39644  36       
39645  37       
39646  38       
39648  39       
39649  40       
39650  41       
39651  42       
39654  43       
39655  44       
39656  45       
39657  46       
39659  47       
39660  48       
39661  49       
39662  50       
39658  51       
39664  52       
39665  53       
39666  54       
39667  55       
39668  56       
39670  57       
39671  58       
39672  59       
39673  60       
39674  61       
38935  62       
39610  63       
39676  64       
39677  65       
39678  66       
39679  67       
39680  68       
39682  69       
39683  70       
39684  71       
39685  72       
39686  73       
39688  74       
39689  75       
39690  76       
39691  77       
39692  78       
39694  79       
39695  80       
39696  81       
39697  82       
39698  83       
39699  84       
39700  85       
39636  86       
39641  87       
39652  88       
39701  89       
39702  90       
39703  91       
39704  92       
39705  93       
39707  94       
39708  95       
39709  96       
39710  97       
39714  98       
39715  99       
39716  100      
39717  101      
39720  102      
39721  103      
39722  104      
39723  105      
39724  106      
39726  107      
39727  108      
39728  109      
39729  110      
39730  111      
39687  112      
39693  113      
39711  114      
39718  115      
39732  116      
39733  117      
39734  118      
39735  119      
39737  120      
39738  121      
39739  122      
39740  123      
39741  124      
39742  125      
39731  126      
39736  127      
39744  128      
39745  129      
39746  130      
39747  131      
39748  132      
39751  133      
39752  134      
39753  135      
39754  136      
39755  137      
39749  138      
39759  139      
39760  140      
39761  141      
39762  142      
39763  143      
39756  144      
39757  145      
39764  146      
39765  147      
39766  148      
39767  149      
39768  150      
39769  151      
39770  152      
39771  153      
39772  154      
39773  155      
39774  156      
39775  157      
39776  158      
39777  159      
39778  160      
39779  161      
39780  162      
39781  163      
39782  164      
39783  165      
39784  166      
39786  167      
39787  168      
39788  169      
39789  170      
39791  171      
39792  172      
39793  173      
39794  174      
39795  175      
39797  176      
39798  177      
39799  178      
39800  179      
39801  180      
39802  181      
39803  182      
39804  183      
39805  184      
39806  185      
39807  186      
39809  187      
39810  188      
39811  189      
39812  190      
39813  191      
39815  192      
39816  193      
39817  194      
39818  195      
39820  196      
39821  197      
39822  198      
39823  199      
39824  200      
39825  201      
39826  202      
39827  203      
39828  204      
39808  205      
39814  206      
39829  207      
39830  208      
39831  209      
39832  210      
39833  211      
39834  212      
39836  213      
39837  214      
39838  215      
39839  216      
39842  217      
39843  218      
39844  219      
39845  220      
39848  221      
39849  222      
39850  223      
39851  224      
39853  225      
39854  226      
39855  227      
39856  228      
39859  229      
39860  230      
39861  231      
39862  232      
39840  233      
39841  234      
39852  235  

In [23]:
# Print the exact line where locations are.
pd.DataFrame(order_article('57928e07cf58f133d1c26609').iloc[179])

39801
sentid  180                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
words   [LOCATION, AND, AGE, OF, FOSSIL, STUDY, SITES, Core, Location, Lat, Long, -LRB-, °, N, -RRB-, -LRB-, °, W, -RRB-, Elevation, Depth, -LRB-, m, -RRB-, -LRB-, cm, -RRB-, Age, Laboratory, -LRB-, yr, B.P., -RRB-, number, Cottonwood, III, 38, °, 49, ′, 50, ′, ′, 106, °, 24, ′, 45, ′, ′, 3670, 40, --, 50, 80, --, 90, 120, --, 130, 2460, ±, 130, 5790, ±, 180, 8230, ±, 200, A-4481, A-4482, A-4483, Cottonwood, IV, 38, °, 49, ′, 50, ′, ′, 106, °, 24, ′, 45, ′, ′, 3670, 60, --, 80, 5610, ±, 150, A-4487, 148, --, 168, 8790, ...]

In [24]:
regex_example = nlp_sentences[nlp_sentences['_gddid'] == '57928e07cf58f133d1c26609']
regex_example = regex_example[regex_example['sentid'] == 180]
regex_example[['words', 'words_as_string', 'dms_regex']]

words  \
39801  [LOCATION, AND, AGE, OF, FOSSIL, STUDY, SITES, Core, Location, Lat, Long, -LRB-, °, N, -RRB-, -LRB-, °, W, -RRB-, Elevation, Depth, -LRB-, m, -RRB-, -LRB-, cm, -RRB-, Age, Laboratory, -LRB-, yr, B.P., -RRB-, number, Cottonwood, III, 38, °, 49, ′, 50, ′, ′, 106, °, 24, ′, 45, ′, ′, 3670, 40, --, 50, 80, --, 90, 120, --, 130, 2460, ±, 130, 5790, ±, 180, 8230, ±, 200, A-4481, A-4482, A-4483, Cottonwood, IV, 38, °, 49, ′, 50, ′, ′, 106, °, 24, ′, 45, ′, ′, 3670, 60, --, 80, 5610, ±, 150, A-4487, 148, --, 168, 8790, ...]   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     words_as_string  \
39801  LOCATION,AND,AGE,OF,FOSSIL,STUDY,SITES,Core,Location,Lat,Long,-LRB-,°,N,-RRB-,-LRB-,°,W,-RRB-,Elevation,Depth,-LRB-,m,-RRB-,-LRB-,cm,-RRB-,Age,Laboratory,-LRB-,yr,B.P.,-RRB-,number,Cottonwood,III,38,°,49,′,50,′,′,106,°,24,′,45,′,′,3670,40,--,50,80,--,90,120,--,130,2460,±,130,5790,±,180,8230,±,200,A-4481,A-4482,A-4483,Cottonwood,IV,38,°,49,′,50,′,′,106,°,24,′,45,′,′,3670,60,--,80,5610,±,150,A-4487,148,--,168,8790,±,230,A-4488,Red,Lady,I,38,°,52,′,50,′,′,107,°,2,′,30,′,′,3350,75,--,95,4675,±,155,GX-8350,Red,Lady,II,38,°,52,′,50,′,′,107,°,2,′,30,′,′,3350,10,--,20,45,--,55,74,--,84,100,±,70,2520,±,90,4680,±,120,A-4484,A-4485,A-4486,Red,Well,38,°,53,′,40,′,′,107,°,3,′,15,′,′,3290,80,--,100,2805,±,160,GX-8351,Copley,II,38,°,52,′,28,′,′,107,°,05,′,3250,25,--,35,62,--,72,102,--,112,125,--,135,155,--,165,181,--,191,212,--,222,1570,±,150,5690,±,150,6900,±,220,8210,±,355,10325,±,465,12810,±,750,15370,±,880,A-4479,A-4480,GX-7906,GX-12995,GX-12996,GX-12997,GX-7907,Splains,II,38,°,50,′,107,°,4,′,30,′,′,3160,45,--,55,90,--,100,190,--,200,266,--,286,291,--,292,3880,±,170,5040,±,140,8750,±,220,11040,±,240,12020,±,530,A-4408,A-4407,A-4409,A-2535,AA-642,Splains,Gulch,I,38,°,50,′,107,°,4,′,30,′,′,3150,34,--,48,80,--,98,134,--,148,190,--,200,4440,±,120,5620,±,140,7340,±,240,8560,±,600,A-4410,A-2534,A-4411,A-2424,Ironbog,38,°,52,′,107,°,2,′,30,′,′,2920,48,--,50,58,--,66,134,--,136,170,195,--,200,218,--,220,2640,±,90,3105,±,155,6300,±,80,7700,±,70,7100,±,255,8260,±,80,AA-1125,GX-7905,AA-1126,AA-1127,GX-7339,AA-1128,Alkali,Basin,I,38,°,45,′,106,°,50,′,2

In [25]:
# Find the article in the bibliography by gddid to retrieve Article Name

bibliography[bibliography['_gddid'] == '57928e07cf58f133d1c26609']

_type                                              _id publisher  \
93  doi   10.1130/0016-7606(1997)109<1306:tfalqp>2.3.co;2  GSA        

                                                                                                  title  \
93  Timberline fluctuations and late Quaternary paleoclimates in the Southern Rocky Mountains, Colorado   

                         journal.name.name             author  year number  \
93  Geological Society of America Bulletin  Fall, Patricia L.  1997  10      

   volumne                    _gddid     type       pages  \
93  NaN     57928e07cf58f133d1c26609  article  1306--1320   

                                                             link_url  \
93  http://dx.doi.org/10.1130/0016-7606(1997)109<1306:tfalqp>2.3.co;2   

    link_type  
93  publisher

## Linking Neotoma DB to extract locations

In [26]:
neotoma = pd.read_csv("../Do_not_commit_data/data-1590729612420.csv")
neotoma = neotoma[['siteid', 'sitename', 'longitudeeast', 'latitudenorth', 'longitudewest', 'latitudesouth', 'sitedescription', 'doi']]

neotoma.head(3)

siteid       sitename  longitudeeast  latitudenorth  longitudewest  \
0  10330   Lac du Sommet -70.66468       47.71662      -70.66573        
1  10330   Lac du Sommet -70.66468       47.71662      -70.66573        
2  1729    Myrtle Lake   -93.37853       47.98645      -93.39207        

   latitudesouth  \
0  47.71382        
1  47.71382        
2  47.97876        

                                                                                                                                                                                                                                                                                                             sitedescription  \
0  The small shallow Lac du Sommet (0.02 km2, 4 m maximum depth, elevation of 830 m a.s.l., 47°43′N, 70°40′W) is located in the boreal forest north of the St Lawrence Estuary (Figure 1). It is of glacial origin and situated on granitic-gneissic bedrock of the Canadian Precambrian Shield in the Laurentian Mountains.   
1  The small shallow Lac du Sommet (0.02 km2, 4 m maximum depth, elevation of 830 m a.s.l., 47°43′N, 70°40′W) is located in the boreal forest north of the St Lawrence Estuary (Figure 1). It is of glacial origin and situated on granitic-gneissic bedrock of the Canadian Precambrian Shield in the Laurentian Mountains.   
2  Lake surrounded by peatland. Physiography: Red Lake lowlands. Surrounding vegetation: Sphagnum, Piceto-Chamaedophnetum.                                                                                                                                                                                                     

                        doi  
0  10.1177/0959683611400199  
1  10.1177/0959683611400199  
2  10.1139/b68-190

In [27]:
nlp_bib = nlp_sentences.merge(bibliography, on = '_gddid')
nlp_bib = nlp_bib.rename(columns={'_id': 'doi'})
nlp_bib.head(3)

_gddid  sentid  \
0  54b43266e138239d8684efed  1        
1  54b43266e138239d8684efed  2        
2  54b43266e138239d8684efed  3        

                                                                                                                                                                                                                                                                                                                                                                                                         wordidx  \
0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, ...]   
1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]                                                                                                                                                                                                                                                                                                                                                                         
2  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]                                                                                                                                                                                                                                                                                                                 

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       words  \
0  [Available, online, at, www.sciencedirect.com, Quaternary, Research, 69, -LRB-, 2008, -RRB-, 263, --, 275, www.elsevier.com/locate/yqres, Development, of, the, mixed, conifer, forest, in, northern, New, Mexico, and, its, relationship, to, Holocene, environmental, change, R., Scott, Anderson, a, ,, b, ,, ,, Renata, B., Jass, b, ,1, ,, Jaime, L., Toney, b, ,2, ,, Craig, D., Allen, c, ,, Luz, M., Cisneros-Dozal, d, ,, Marcey, Hess, d, ,, Jeff, Heikoop, d, ,, Julianna, Fessenden, d, a, Center, for, Environmental, Sciences, &, Education, ,, Box, 5694, ,, Northern, Arizona, University, ,, Flagstaff, ,, AZ, 86011, ,, USA, b, Quaternary, Sciences, Program, &, Bilby, Research, ...]   
1  [The, Chihuahueños, Bog, record, extends, to, over, 15,000, cal, yr, BP, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
2  [An, Artemisia, steppe, ,, then, an, open, Picea, woodland, grew, around, a, small, pond, until, ca., 11,700, cal, yr, BP, when, Pinus, ponderosa, became, established, .]          

In [28]:
print(nlp_bib.shape)

(111965, 24)


In [29]:
nlp_bib[['words_as_string', 'journal.name.name', 'dms_regex']].head(3)

words_as_string  \
0  Available,online,at,www.sciencedirect.com,Quaternary,Research,69,-LRB-,2008,-RRB-,263,--,275,www.elsevier.com/locate/yqres,Development,of,the,mixed,conifer,forest,in,northern,New,Mexico,and,its,relationship,to,Holocene,environmental,change,R.,Scott,Anderson,a,,,b,,,,,Renata,B.,Jass,b,,1,,,Jaime,L.,Toney,b,,2,,,Craig,D.,Allen,c,,,Luz,M.,Cisneros-Dozal,d,,,Marcey,Hess,d,,,Jeff,Heikoop,d,,,Julianna,Fessenden,d,a,Center,for,Environmental,Sciences,&,Education,,,Box,5694,,,Northern,Arizona,University,,,Flagstaff,,,AZ,86011,,,USA,b,Quaternary,Sciences,Program,&,Bilby,Research,Center,,,Box,6013,,,Northern,Arizona,University,,,Flagstaff,,,AZ,86011,,,USA,c,U.S.,Geological,Survey,,,Jemez,Mountains,Field,Station,,,Bandelier,National,Monument,,,HCR-1,,,Box,1,,,#,15,,,Los,Alamos,,,NM,87544,,,USA,d,Earth,&,Environmental,Sciences,Division,,,Hydrology,,,Geochemistry,&,Geology,Group,,,EES-6,,,MS-D462,,,Los,Alamos,National,Laboratory,,,Los,Alamos,,,NM,87545,,,USA,Received,28,April,2007,Available,online,29,January,2008,Abstract,Chihuahueños,Bog,-LRB-,2925,m,-RRB-,in,the,Jemez,Mountains,of,northern,New,Mexico,contains,one,of,the,few,records,of,late-glacial,and,postglacial,development,of,the,mixed,conifer,forest,in,southwestern,North,America,.   
1  The,Chihuahueños,Bog,record,extends,to,over,15,000,cal,yr,BP,.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
2  An,Artemisia,steppe,,,then,an,open,Picea,woodland,grew,around,a,small,pond,until,ca.,11,700,cal,yr,BP,when,Pinus,ponderosa,became,established,.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             

     journal.name.name dms_regex  
0  Quaternary Research  []        
1  Quaternary Research  []        
2  Quaternary Research  []

Let's now try to extract Sitenames using Neotoma database. For this, I will group by doi and then, list all sitenames that are mentioned in each doi.

In [30]:
by_doi = neotoma.groupby('doi')
neotoma_sitenames = by_doi.apply(lambda x: list(x['sitename']))
neotoma_siteid = by_doi.apply(lambda x: list(x['siteid']))
neotoma_longeast = by_doi.apply(lambda x: list(x['longitudeeast']))
neotoma_latnorth = by_doi.apply(lambda x: list(x['latitudenorth']))

neotoma_joined_df = neotoma_sitenames.to_frame('sitenames').join(neotoma_siteid.to_frame('siteid')).join(neotoma_longeast.to_frame('longeast')).join(neotoma_latnorth.to_frame('latnorth'))
neotoma_joined_df = neotoma_joined_df.reset_index()

#neotoma_joined_df['intlongeast'] = [int(i) for neotoma_joined_df['longeast'][i]]

#print(neotoma_joined_df['longeast'][1].astype(int))
#[round(x) for x in myList]
neotoma_joined_df.head(5)


doi  \
0  0.1139/e80-122                                                      
1  10.1002/(SICI)1099-1417(199607/08)11:4<327::AID-JQS257>3.0.CO;2-G   
2  10.1002/(SICI)1099-1417(199701/02)12:1<15::AID-JQS283>3.0.CO;2-T    
3  10.1002/(SICI)1099-1417(199801/02)13:1<43::AID-JQS348>3.0.CO;2-G    
4  10.1002/1099-1417(200010)15:7<687::AID-JQS551>3.0.CO;2-6            

                                     sitenames              siteid  \
0  [East Baltic Bog, Mermaid Bog, Portage Bog]  [763, 1684, 1903]    
1  [Lundin Tower]                               [3256]               
2  [Spruce Pond, Spruce Pond, Spruce Pond]      [2508, 2508, 2508]   
3  [Pantano de Genagra, Pantano de Genagra]     [14030, 14030]       
4  [Laguna Loma Linda, Laguna Loma Linda]       [13971, 13971]       

                            longeast  \
0  [-62.15, -63.025, -64.075]          
1  [-2.97206]                          
2  [-74.18236, -74.18236, -74.18236]   
3  [-76.60428, -76.60428]              
4  [-73.35661, -73.35661]              

                                                       latnorth  
0  [46.408333, 46.25, 46.673611]                                 
1  [56.21529]                                                    
2  [41.237770000000005, 41.237770000000005, 41.237770000000005]  
3  [2.4751, 2.4751]                                              
4  [3.30805, 3.30805]

From Neotoma DB, we have obtained a dataframe grouped by DOI that lists each `sitename`, `siteid`, `longitude east`, and `latitunde north`. Now, let's join this information to the NLP_Bib dataframe. For now, let's do to the one that coordinates have been obtained with REGEX just to see if some of the numbers and sitenames make sense. 

In [31]:
nlp_bib_neotoma = nlp_bib.merge(neotoma_joined_df, on = 'doi')
print(nlp_bib_neotoma.shape)

(106640, 28)


In [32]:
# Filter df to just frames that has REGEX longer than one
mask = (nlp_bib_neotoma['dms_regex'].str.len() > 0)
nlp_bib_neotoma_regex = nlp_bib_neotoma.loc[mask]
nlp_bib_neotoma_regex[['words','dms_regex']].head(100)

print(nlp_bib_neotoma_regex.shape)

(93, 28)


In [33]:
nlp_bib_neotoma_regex = nlp_bib_neotoma_regex[['_gddid', 'sentid', 'doi', 'words', 'words_as_string', 'dms_regex', 'sitenames', 'latnorth', 'longeast']]
nlp_bib_neotoma_regex = nlp_bib_neotoma_regex.reset_index()
nlp_bib_neotoma_regex.head(1)

index                    _gddid  sentid              doi  \
0  3314   578a1f6ecf58f1b54b810cd9  94      10.1139/b85-070   

                                                                                                                                                                                                                                               words  \
0  [Lake, Hope, Simpson, -LRB-, 52, ', 27, ', N, ,, 56, ', 26, ', W, ,, elevation, 295, m, -RRB-, lies, 20, km, southwest, of, Port, Hope, Simpson, in, a, steep-sided, eastwest-trending, valley, that, is, drained, by, the, St., Lewis, River, .]   

                                                                                                                                                                                          words_as_string  \
0  Lake,Hope,Simpson,-LRB-,52,',27,',N,,,56,',26,',W,,,elevation,295,m,-RRB-,lies,20,km,southwest,of,Port,Hope,Simpson,in,a,steep-sided,eastwest-trending,valley,that,is,drained,by,the,St.,Lewis,River,.   

                     dms_regex                               sitenames  \
0  [52,',27,',N,,,56,',26,',W]  [Lake Hope Simpson, 18 [Moraine Lake]]   

             latnorth              longeast  
0  [52.45, 52.266667]  [-56.433333, -58.05]

In [34]:
any( k in s for k in keywords )

NameError: name 'keywords' is not defined

We can see that most of the REGEX indeed pulled out the same locations for latitude and longitudes as in the Neotoma database.

In [35]:
nlp_bib_neotoma[['_gddid', 'sentid', 'doi', 'words', 'dms_regex', 'sitenames', 'latnorth', 'longeast']].head(2)

_gddid  sentid                          doi  \
0  54b43266e138239d8684efed  1       10.1016/j.yqres.2007.12.002   
1  54b43266e138239d8684efed  2       10.1016/j.yqres.2007.12.002   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       words  \
0  [Available, online, at, www.sciencedirect.com, Quaternary, Research, 69, -LRB-, 2008, -RRB-, 263, --, 275, www.elsevier.com/locate/yqres, Development, of, the, mixed, conifer, forest, in, northern, New, Mexico, and, its, relationship, to, Holocene, environmental, change, R., Scott, Anderson, a, ,, b, ,, ,, Renata, B., Jass, b, ,1, ,, Jaime, L., Toney, b, ,2, ,, Craig, D., Allen, c, ,, Luz, M., Cisneros-Dozal, d, ,, Marcey, Hess, d, ,, Jeff, Heikoop, d, ,, Julianna, Fessenden, d, a, Center, for, Environmental, Sciences, &, Education, ,, Box, 5694, ,, Northern, Arizona, University, ,, Flagstaff, ,, AZ, 86011, ,, USA, b, Quaternary, Sciences, Program, &, Bilby, Research, ...]   
1  [The, Chihuahueños, Bog, record, extends, to, over, 15,000, cal, yr, BP, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

  dms_regex                             sitenames            latnorth  \
0  []        [Chihuahueños Bog, Chihuahueños Bog]  [36.0491, 36.0491]   
1  []        [Chihuahueños Bog, Chihuahueños Bog]  [36.0491, 36.0491]   

                   longeast  
0  [-106.50963, -106.50963]  
1  [-106.50963, -106.50963]

In [36]:
#nlp_bib_neotoma['words_l'] = nlp_bib_neotoma['words'].apply(lambda col: col.str.lower())

nlp_bib_neotoma['words_l']  = [s.lower() for s in nlp_bib_neotoma['words']]
nlp_bib_neotoma[['words', 'words_l']].head(2)



AttributeError: 'list' object has no attribute 'lower'

In [58]:
a = nlp_bib_neotoma_regex['sitenames'][48]
a = [x.upper() for x in a]

In [59]:
b = nlp_bib_neotoma_regex['words'][48]

In [60]:
list(set(a).intersection(b))

['UDDELERMEER', 'BLEEKEMEER']

In [56]:
print(type(nlp_bib_neotoma['sitenames'][48]))
print(type(nlp_bib_neotoma['words'][48]))
print(type(nlp_bib_neotoma['words_as_string'][48]))

<class 'list'>
<class 'list'>
<class 'str'>


In [43]:
#for i in range(len(nlp_bib_neotoma.index)):
#    nlp_bib_neotoma['sitename_in_wordsstr'] = any(item in nlp_bib_neotoma['words_as_string'][i] for item in nlp_bib_neotoma['sitenames'][i])
    
    
    

In [265]:
#nlp_bib_neotoma.head(299)